homeassistant/components/zwave_mqtt/__init__.py

```diff
@@ -0,0 +1,333 @@
+"""The zwave_mqtt integration."""
+import asyncio
+import json
+import logging
+
+from openzwavemqtt import OZWManager, OZWOptions
+from openzwavemqtt.const import (
+    EVENT_INSTANCE_EVENT,
+    EVENT_NODE_ADDED,
+    EVENT_NODE_CHANGED,
+    EVENT_NODE_REMOVED,
+    EVENT_VALUE_ADDED,
+    EVENT_VALUE_CHANGED,
+    EVENT_VALUE_REMOVED,
+    CommandClass,
+    ValueType,
+)
+from openzwavemqtt.models.node import OZWNode
+from openzwavemqtt.models.value import OZWValue
+import voluptuous as vol
+
+from homeassistant.components import mqtt
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.device_registry import async_get_registry as get_dev_reg
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+
+from . import const
+from .const import DATA_UNSUBSCRIBE, DOMAIN, PLATFORMS, TOPIC_OPENZWAVE
+from .discovery import DISCOVERY_SCHEMAS, check_node_schema, check_value_schema
+from .entity import ZWaveDeviceEntityValues, create_device_id
+from .services import ZWaveServices
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+DATA_DEVICES = "zwave-mqtt-devices"
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Initialize basic config of zwave_mqtt component."""
+    if "mqtt" not in hass.config.components:
+        _LOGGER.error("MQTT integration is not set up")
+        return False
+    hass.data[DOMAIN] = {}
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up zwave_mqtt from a config entry."""
+
+    @callback
+    def async_receive_message(msg):
+        manager.receive_message(msg.topic, msg.payload)
+
+    platforms_loaded = []
+
+    async def mark_platform_loaded(platform):
+        platforms_loaded.append(platform)
+
+        if len(platforms_loaded) != len(PLATFORMS):
+            return
+
+        hass.data[DOMAIN][entry.entry_id][DATA_UNSUBSCRIBE].append(
+            await mqtt.async_subscribe(
+                hass, f"{TOPIC_OPENZWAVE}/#", async_receive_message
+            )
+        )
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        "mark_platform_loaded": mark_platform_loaded,
```

Comment:
Just saw an alternative approach to this in #34447 which I like more:

![image](https://user-images.githubusercontent.com/1444314/80770259-a4d87280-8b04-11ea-8014-760383212491.png)

Reply:
Nice! But don't pass the loop.

Reply:
haha yes, and it shouldn't create a list of known entries and then splat it, it should just pass as positional args.

Reply:
Commit done.